### PR TITLE
Generate Haskell datatype declarations from Dhall types

### PR DESCRIPTION
**Pull request by Gabriel439**
Fixes https://github.com/dhall-lang/dhall-haskell/issues/1616

This adds a new `Dhall.TH.makeHaskellType` utility which generates a
Haskell datatype declaration corresponding to a Dhall type.  This simplifies
keeping Haskell and Dhall code in sync with one another.

### CODE REVIEW

File: dhall/src/Dhall/TH.hs [hunk 1]
```diff
@@ -22,28 +66,209 @@
...
+-- >
+-- > Dhall.TH.makeHaskellType "T" "< A : { x : Bool } | B >"
+-- > 
+-- > deriving instance Generic   T
+-- > deriving instance FromDhall T
+--
+-- Carefully note that the generated Haskell type only corresponds to the Dhall
+-- type when using the `Dhall.Smart` setting for the
+-- `Dhall.singletonConstructors` field of `InterpretOptions`.  That setting
+-- will eventually become the default, but until then you will need to
```
Comment (sjakobi) [inline dhall/src/Dhall/TH.hs]: How about just changing the default right now, to simplify these docs?!

File: dhall/src/Dhall/TH.hs [hunk 2]
```diff
@@ -22,28 +66,209 @@
...
+-- > let decoder = Dhall.autoWith interpretOptions
+
+makeHaskellType :: Text -> Text -> Q [Dec]
+makeHaskellType typeName text = do
+    Syntax.runIO (GHC.IO.Encoding.setLocaleEncoding System.IO.utf8)
+
+    expression <- Syntax.runIO (Dhall.inputExpr text)
+
+    case expression of
+        Union kts -> do
```
Comment (sjakobi) [inline dhall/src/Dhall/TH.hs]: Is the restriction to unions a design limitation, or is this just a first step and support for records etc could be added?

File: nix/shared.nix [hunk 3]
```diff
@@ -399,6 +399,11 @@ let
                           haskellPackagesNew.semigroups
                         ];
 
+                    managed =
+                      pkgsNew.haskell.lib.addBuildDepend
+                        haskellPackagesOld.managed
+                        haskellPackagesNew.semigroups;
```
Comment (sjakobi) [inline nix/shared.nix]: How is this change related to the PR?

Comment (Gabriel439) [reply nix/shared.nix]: It's not.  It's a small change I made to get `nix-shell` working for GHC 7.10.3 when debugging a build failure

File: dhall/src/Dhall/TH.hs (hunk 1 above)
Comment (Gabriel439) [reply dhall/src/Dhall/TH.hs]: Good idea!  I'll go ahead and change the default

File: dhall/src/Dhall/TH.hs (hunk 2 above)
Comment (Gabriel439) [reply dhall/src/Dhall/TH.hs]: @sjakobi: It is possible to support records, but it would have to be via a separate Template Haskell function.  The reason why is that this function has no way to infer what to name the constructor wrapping the record if there isn't a union wrapping the record.  If there were a separate record-specific function it could take the constructor name as an additional argument.

Comment (sjakobi) [reply dhall/src/Dhall/TH.hs]: If this function will only work for unions, how about calling it sth like `makeHaskellUnionType` or `makeHaskellTypeForUnion`?

Comment (Gabriel439) [reply dhall/src/Dhall/TH.hs]: Actually, now that I think about it, there's no benefit to having a separate function for bare records.  Instead of doing this:

```haskell
makeHaskellTypeFromRecord "T" "C" "./some/record.dhall"
```

... they could always do:

```haskell
makeHaskellTypeFromUnion "T" "< C : ./some/record.dhall >"
```

However, I could update the documentation and error message to explain that
